### PR TITLE
DB setup-teardown compatible with tests

### PR DESCRIPTION
**Pull request by gyermolenko**
Hi all. I'm stuck and need some advice.
Trying to implement db setup/teardown process compatible with tests.

I want to have separate db created exclusively for specified test and removed afterwards.
Is it possible to do so asynchronously? How two async tests can manipulate db without failing each other?

So far I kill all sessions in teardown_db - and because of that can run only one test in time.

### CODE REVIEW

**Comment by webknjaz** [comment]
You don't need to do this asynchronously. Tests are executed sequentially within a single pytest process anyway. Just use `engine.create_all()` (from SQLAlchemy) and that's it!

**Comment by gyermolenko** [comment]
@webknjaz hi. Thanks for reply.
I think you meant meta.create_all() - I use exactly that in create_tables method. 

Also, I can't see reasons why tests can't be run in parallel (and/or async). Built-in unittest module can do that, nose-tests can do that...

**Comment by webknjaz** [comment]
pytest has pytest-xdist for parallel execution, but that spawns separate processes AFAIK

**Comment by gyermolenko** [comment]
To clarify my situation here:
What I wanted (silly me) is the following scenario:
test1:  no test db -> test db with prepared data -> 1 of tests -> no test db.
test2:  no test db -> test db with prepared data -> 1 of tests -> no test db.
etc
 
But having two or more tests: one of them creates db, another tries to do the same, first one destroys everything, second one fails. Something like that.

So here I rather wanted to find out best practices for such cases.
I think I will try something like [this](https://docs.djangoproject.com/en/1.11/topics/testing/overview/#the-test-database) . 
Which is:
no db -> empty test db with no tables ->
test1:  test db with prepared data -> test -> empty test db with no tables 
test2:  test db with prepared data -> test -> empty test db with no tables 
no db after

**Comment by webknjaz** [comment]
@gyermolenko I think you're seeking for `@pytest.fixture(scope='module')`, which would run once per test module (you could create a db this way). And then you could do regular per-function fixture filing in the DB with test data and then removing it.

**Comment by gyermolenko** [comment]
1. @webknjaz thanks, `@pytest.fixture(scope='module')` did the trick.
I changed logic to create db once per test run and tables creation/drop once per test function.

2. Fixed test for posting a choice.

3. Currently there is config duplication in `polls.yml` and new `config.py` with simple config dicts. If everybody is ok with DB initialization approach from this PR - I will remove sql/sh/yml files from `polls` app. 
Please review `init_db.py` keeping in mind that it should serve two purposes:
   - application setup
   - tests

**Comment by webknjaz** [comment]
JFYI `scope='session'` is once per test run and `scope='module'` is once per python module.

File: demos/polls/aiohttpdemo_polls/config.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+#
+# minsize: 1
+# maxsize: 5
+#
+# app_host = '127.0.0.1'
+# app_port = '8080'
+
+
+db_host = 'localhost'
```
Comment (jettify) [inline demos/polls/aiohttpdemo_polls/config.py]: can we move this values to yaml file?

File: demos/polls/tests/init_db.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
+from sqlalchemy import create_engine, MetaData
+from aiohttpdemo_polls.db import question, choice
+
+from aiohttpdemo_polls.config import DB_CONFIG_ADMIN, DB_CONFIG_USER
+
+
+DSN = "postgresql://{user}:{password}@{host}:{port}/{database}"
+
+DB_URL = DSN.format(**DB_CONFIG_ADMIN)
+engine = create_engine(DB_URL, isolation_level='AUTOCOMMIT')
```
Comment (jettify) [inline demos/polls/tests/init_db.py]: I suggest to wrap engine into separate fixture

**Comment by jettify** [comment]
I think it is good start
1) config should be cleaned and yaml file used instead
2) Engine creation should be wrapped into separate fixture

and we can merge PR. We can iterate and break things quickly so no point to be perfect in everything

**Comment by webknjaz** [comment]
I 100% agree with @jettify and would only add that delegating config processing to smth like `aio_yamlconfig` would make things cleaner. But at least plz drop config from code in favour of structured YAML.

**Comment by gyermolenko** [comment]
@jettify @webknjaz Hi. Appreciate your comments. Don't you think trafaret_config / aio_yamlconfig is an overkill for demo? I think that simple dict would do here just fine.

**Comment by webknjaz** [comment]
@gyermolenko people tend to repeat what you show them. It's better to demonstrate best practices from the beginning.

**Comment by jettify** [comment]
not overkill, having config in python file, in my practice, leads to disaster.

**Comment by jettify** [comment]
@gyermolenko there are flake8 errors, can you please fix them in separate PR, should be very easy? Not big problem or rush we just need to address this at some point.
```
demos/polls/tests/init_db.py:12:1: E302 expected 2 blank lines, found 1
demos/polls/tests/init_db.py:44:80: E501 line too long (82 > 79 characters)
demos/polls/tests/init_db.py:68:1: E302 expected 2 blank lines, found 1
demos/polls/tests/init_db.py:76:80: E501 line too long (83 > 79 characters)
demos/polls/tests/test_integration.py:3:1: F401 'aiohttpdemo_polls.db.question' imported but unused
demos/polls/tests/test_integration.py:26:17: E131 continuation line unaligned for hanging indent
demos/polls/tests/test_integration.py:41:17: E131 continuation line unaligned for hanging indent
```

**Comment by gyermolenko** [comment]
@jettify sure

**Pull request merged**